examples/xlio_ultra_api_ping_pong.c

```diff
@@ -423,74 +290,189 @@ static int create_polling_group(void)
         }
-    } else {
-        addr.sin_addr.s_addr = INADDR_ANY;
-    }
 
-    rc = g_state.api->xlio_socket_bind(g_state.sock, (struct sockaddr *)&addr, sizeof(addr));
-    if (rc != 0) {
-        fprintf(stderr, "Failed to bind server socket: %s\n", strerror(errno));
-        return rc;
+        /* Wait for termination event */
+        while (!g_state.terminated) {
+            g_state.api->xlio_poll_group_poll(g_state.group);
+        }
     }
 
-    printf("Server socket bound to %s:%u\n", g_config.ip ? g_config.ip : "*", g_config.port);
-
-    /*
-     * Step 3: Start listening for connections
-     * =======================================
-     * Note: socket_accept_cb must be registered in the polling group for this to work.
-     */
+    /* Destroy main socket */
+    if (g_state.sock) {
+        g_state.terminated = false;
+        rc = g_state.api->xlio_socket_destroy(g_state.sock);
+        if (rc != 0) {
+            fprintf(stderr, "Failed to destroy socket\n");
+        }
+
+        /* Wait for termination event */
+        while (!g_state.terminated) {
+            g_state.api->xlio_poll_group_poll(g_state.group);
+        }
+    }
+
+    /* Destroy polling group */
+    if (g_state.group) {
+        rc = g_state.api->xlio_poll_group_destroy(g_state.group);
+        if (rc != 0) {
+            fprintf(stderr, "Failed to destroy polling group\n");
+        }
+    }
+
+    /* Cleanup RX accumulator */
+    if (g_state.rx_acc.data) {
+        free(g_state.rx_acc.data);
+        g_state.rx_acc.data = NULL;
+    }
+
+    /* Deregister and free memory for TX buffer pool */
+    cleanup_tx_buffer_pool();
+
+    /* Finalize XLIO */
+    rc = g_state.api->xlio_exit();
+    if (rc != 0) {
+        fprintf(stderr, "Failed to finalize XLIO\n");
+    }
+
+    printf("Resources cleaned up\n");
+}
+
+/*
+ * Send message using zero-copy operation with buffer pool
+ * =======================================================
+ * Performs the following steps:
+ *  1. Gets a buffer from the TX buffer pool
+ *  2. Copies the message
+ *  3. Adds a newline delimiter according to our protocol
+ *  4. The buffer pointer is stored in userdata_op for identification in completion callback
+ *  5. Sends using zero-copy interface
+ */
+static int send_message(xlio_socket_t sock, const char *msg)
+{
+    size_t msg_len = strlen(msg);
+    size_t total_len = msg_len + 1; /* +1 for '\n' */
+    char *buffer;
+
+    if (!g_state.mr_buf) {
+        fprintf(stderr, "TX buffer pool not initialized\n");
+        return -1;
+    }
+
+    if (total_len >= TX_BUFFER_SIZE) {
+        fprintf(stderr, "Message too long for send buffer (max %d bytes)\n", TX_BUFFER_SIZE - 1);
+        return -1;
+    }
+
+    /* Get buffer from TX buffer pool */
+    buffer = get_tx_buffer();
+    if (!buffer) {
+        fprintf(stderr, "No available TX buffers in pool\n");
+        return -1;
+    }
+
+    /* Construct TCP payload according to our protocol */
```

Comment:
this is app-level message construction, not related to TCP. change to:
/* Construct application message with newline delimiter */